src/test/java/co/cask/cdc/plugins/integration/sink/CDCBigTablePluginTest.java

```diff
@@ -0,0 +1,387 @@
+import co.cask.cdc.plugins.sink.CDCBigTableConfig;
+import co.cask.cdc.plugins.sink.CDCTableUtil;
+import co.cask.hydrator.common.Constants;
+import com.google.bigtable.repackaged.com.google.cloud.ServiceOptions;
+import com.google.bigtable.repackaged.io.grpc.StatusRuntimeException;
+import com.google.cloud.bigtable.hbase.BigtableConfiguration;
+import com.google.cloud.bigtable.hbase.BigtableOptionsFactory;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableMap;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.Connection;
+import org.apache.hadoop.hbase.client.Get;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.Table;
+import org.assertj.core.api.Assertions;
+import org.awaitility.Awaitility;
+import org.awaitility.Duration;
+import org.junit.After;
+import org.junit.Assume;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.runners.Enclosed;
+import org.junit.rules.TestName;
+import org.junit.runner.RunWith;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@RunWith(Enclosed.class)
+public class CDCBigTablePluginTest {
+  private static final String PLUGIN_NAME = "CDCBigTable";
+  private static final String APP_NAME = CDCBigTablePluginTest.class.getSimpleName();
+  private static final String PROJECT
+    = System.getProperty("test.bigtable.project", ServiceOptions.getDefaultProjectId());
+  private static final String INSTANCE = System.getProperty("test.bigtable.instance");
+  private static final String SERVICE_ACCOUNT_FILE_PATH
+    = System.getProperty("test.bigtable.serviceFilePath", System.getenv("CREDENTIAL_ENV_NAME"));
+  private static final String DB_NAMESPACE = "dbNamespace";
+
+  private static void dropTableIfExists(Connection connection, String dbTableName) throws IOException {
+    TableName tableName = TableName.valueOf(dbTableName);
+    if (connection.getAdmin().tableExists(tableName)) {
+      connection.getAdmin().disableTable(tableName);
+      connection.getAdmin().deleteTable(tableName);
+    }
+  }
+
+  private static byte[] getColumnValue(Result result, String column) {
+    return result.getValue(Bytes.toBytes(CDCTableUtil.CDC_COLUMN_FAMILY), Bytes.toBytes(column));
+  }
+
+  public static class Validations extends CDCPluginTestBase {
+    @Rule
+    public TestName testName = new TestName();
+
+    @Before
+    public void beforeMethod() {
+      Assume.assumeNotNull(PROJECT);
+      Assume.assumeNotNull(INSTANCE);
+    }
+
+    @Test
+    public void testDeploymentFailedWhenCredentialsFileIsMissing() throws Exception {
+      ETLPlugin sourceConfig = MockSource.getPlugin(Schemas.CHANGE_SCHEMA, Collections.emptyList());
```

Comment:
There is no need to test this through pipeline deployment. It is much less expensive to test these things in a separate CDCBigTableConfigTest class where you create instances of CDCBigTableConfig and just call validate().

Reply:
I did it for several reasons:
- to make tests look similar because we will still need config tests with real pipeline. Splitting those tests will make it more difficult to understand what exactly is covered by tests.
- to make future refactoring easier. As for me "validate" methods inside configuration DTO object violates single responsibility principle a bit.
- to test mappings (if everything is correctly mapped from properties map to config instance)

What do you think?

Reply:
The only plugin specific logic that is being tested here is the validation logic in the config. Everything else is functionality that CDAP is supposed to provide, and is unit tested in the CDAP project. When possible, we just want to test plugin logic, not to test plugin logic while also testing all the stuff CDAP does.